engine/src/main/java/com/agiletec/aps/system/services/i18n/cache/I18nManagerCacheWrapper.java

```diff
@@ -0,0 +1,117 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cache.Cache;
+
+import com.agiletec.aps.system.common.AbstractCacheWrapper;
+import com.agiletec.aps.system.exception.ApsSystemException;
+import com.agiletec.aps.system.services.i18n.II18nDAO;
+import com.agiletec.aps.util.ApsProperties;
+
+public class I18nManagerCacheWrapper extends AbstractCacheWrapper implements II18nManagerCacheWrapper {
+
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+
+	@Override
+	protected String getCacheName() {
+		return I18N_MANAGER_CACHE_NAME;
+	}
+
+	@Override
+	public void initCache(II18nDAO i18nDAO) throws ApsSystemException {
+		try {
+			Cache cache = this.getCache();
+			this.releaseCachedObjects(cache);
+			Map<String, ApsProperties> labels = i18nDAO.loadLabelGroups();
+			this.insertObjectsOnCache(cache, labels);
+		} catch (Throwable t) {
+			logger.error("Error loading labels", t);
+			throw new ApsSystemException("Error loading labels", t);
+		}
+	}
+
+	private void insertObjectsOnCache(Cache cache, Map<String, ApsProperties> labels) {
+		for (Map.Entry<String, ApsProperties> entry : labels.entrySet()) {
+			String key = entry.getKey();
+			cache.put(I18N_CACHE_NAME_PREFIX + key, entry.getValue());
+		}
+		cache.put(I18N_CODES_CACHE_NAME, labels);
+	}
+
+	protected void releaseCachedObjects(Cache cache) {
+		this.releaseCachedObjects(cache, I18N_CODES_CACHE_NAME, I18N_CACHE_NAME_PREFIX);
+	}
+
+	@SuppressWarnings("unchecked")
+	protected void releaseCachedObjects(Cache cache, String cacheKey, String codePrefix) {
+		Map<String, ApsProperties> groupCodes = (Map<String, ApsProperties>) this.get(cache, cacheKey, Map.class);
+		if (null != groupCodes) {
+			for (Map.Entry<String, ApsProperties> entry : groupCodes.entrySet()) {
+				String key = entry.getKey();
+				cache.evict(codePrefix + key);
+			}
+			cache.evict(cacheKey);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public Map<String, ApsProperties> getLabelGroups() {
+		return this.get(this.getCache(), I18N_CODES_CACHE_NAME, Map.class);
+	}
+
+	@Override
+	public void addLabelGroup(String key, ApsProperties labels) {
+		this.manage(key, labels, Action.ADD);
+	}
+
+	@Override
+	public void updateLabelGroup(String key, ApsProperties labels) {
+		this.manage(key, labels, Action.UPDATE);
+	}
+
+	@Override
+	public void removeLabelGroup(String key) {
+		this.manage(key, null, Action.DELETE);
+	}
+
+	private void manage(String key, ApsProperties labels, Action operation) {
+		if (StringUtils.isBlank(key)) {
+			return;
+		}
+		Cache cache = this.getCache();
+		Map<String, ApsProperties> codes = this.getLabelGroups();
+		if (Action.ADD.equals(operation)) {
+			if (!codes.containsKey(key)) {
+				codes.put(key, labels);
+				cache.put(I18N_CODES_CACHE_NAME, codes);
+			}
+			cache.put(I18N_CACHE_NAME_PREFIX + key, labels);
+		} else if (Action.UPDATE.equals(operation) && codes.containsKey(key)) {
```

Comment:
What happens if UPDATE is called but the corresponding key isn't in the cache? Right now this is a no-op which could be really confusing and could point to a cache consistency issue in a race between an update and a delete. This should probably throw an exception so that we can return to a user that the object they are updating no longer exists. Or it should behave as an UPSERT which has its own issues. Whatever the case it will need to be consistent with the other implementations.